### PR TITLE
ticdc/owner: Fix ddl special comment syntax error

### PR DESCRIPTION
**Pull request by sdojjy**
### What problem does this PR solve? 
close https://github.com/pingcap/ticdc/issues/3755

### What is changed and how it works?
not use regex to parse and add special TiDB comments,  If a DDL already has special comment like `/*!90000  SHARD_ROW_ID_BITS=4 PRE_SPLIT_REGIONS=3 */`, it wil translate that DDL to `/*!90000 /*T! SHARD_ROW_ID_BITS=4 PRE_SPLIT_REGIONS=3 */ */`, TiDB will throw a syntax error.
this PR use  the TiDB SQL parser to parse the SQL and the Restore it with `format.RestoreStringSingleQuotes|format.RestoreNameBackQuotes|format.RestoreKeyWordUppercase|format.RestoreTiDBSpecialComment` flag

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update key monitor metrics in both TiCDC document and official document

### Release note 

```release-note
Fix syntax error if DDL has a special comment.
```

### CODE REVIEW

File: cdc/owner/changefeed.go [hunk 1]
```diff
@@ -494,3 +500,19 @@ func (c *changefeed) updateStatus(currentTs int64, checkpointTs, resolvedTs mode
 func (c *changefeed) Close(ctx context.Context) {
 	c.releaseResources(ctx)
 }
+
+// addSpecialComment translate tidb feature to comment
+func addSpecialComment(ddlQuery string) (string, error) {
```
Comment (overvenus) [inline cdc/owner/changefeed.go:505]: Are special comments compatible with older version of TiDB (v4.0.0)?

**Review by overvenus — COMMENTED** [review]
Rest LGTM

**Comment by overvenus** [comment]
Please also add a release note, thanks!

File: cdc/owner/changefeed.go [hunk 2]
```diff
@@ -528,3 +534,20 @@ func (c *changefeed) updateStatus(currentTs int64, checkpointTs, resolvedTs mode
...
+	stms, _, err := parser.New().ParseSQL(ddlQuery)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	if len(stms) != 1 {
+		log.Panic("invalid ddlQuery statement size", zap.String("ddlQuery", ddlQuery))
+	}
+	var sb strings.Builder
+	if err = stms[0].Restore(format.NewRestoreCtx(format.RestoreStringSingleQuotes|format.RestoreNameBackQuotes|
+		format.RestoreKeyWordUppercase|format.RestoreTiDBSpecialComment, &sb)); err != nil {
```
Comment (overvenus) [inline cdc/owner/changefeed.go:549]: Could you comment these flag usage correspondingly?

Also, I'd prefer the following style:

```go
// Comments ...
restoreFlags := RestoreStringSingleQuotes
// Comments ...
restoreFlags |= RestoreNameBackQuotes
...
```

**Review by overvenus — COMMENTED** [review]
Rest LGTM

File: cdc/owner/changefeed.go (hunk 2 above)
Comment (sdojjy) [reply cdc/owner/changefeed.go]: done

**Review by amyangfei — APPROVED** [review]

**Comment by sdojjy** [comment]
# Test below DDLs on TiDB
```sql
create database if not exists ddl_test;
use ddl_test;
CREATE TABLE `t1` (`id` INT) /*T! SHARD_ROW_ID_BITS = 2 */ /*T! SHARD_ROW_ID_BITS = 3 */ /*T! PRE_SPLIT_REGIONS = 2 */;
CREATE TABLE `t2` (`id` INT /*T![auto_rand] AUTO_RANDOM(3) */ PRIMARY KEY) /*T![auto_rand_base] AUTO_RANDOM_BASE = 100 */;
CREATE TABLE `t3` (`id` INT /*T![auto_rand] AUTO_RANDOM */ PRIMARY KEY) /*T![auto_rand_base] AUTO_RANDOM_BASE = 50 */;
CREATE TABLE `t4` (`id` INT AUTO_INCREMENT PRIMARY KEY) /*T![auto_id_cache] AUTO_ID_CACHE = 100 */;
CREATE TABLE `t5` (`id` INT AUTO_INCREMENT UNIQUE KEY) /*T![auto_id_cache] AUTO_ID_CACHE = 10 */;
CREATE TABLE `t6` (`id` INT) /*T![auto_id_cache] AUTO_ID_CACHE = 5 */;
CREATE TABLE `t7` (`a` INT,`b` VARCHAR(255),PRIMARY KEY(`a`, `b`) /*T![clustered_index] CLUSTERED */);
CREATE TABLE `t8` (`a` INT,`v` INT,PRIMARY KEY(`a`) /*T![clustered_index] CLUSTERED */);
CREATE TABLE `t9` (`id` INT PRIMARY KEY /*T![clustered_index] CLUSTERED */,`v` INT);
CREATE TABLE `t10` (`id` INT);
ALTER TABLE `t10` ADD PRIMARY KEY(`id`) /*T![clustered_index] CLUSTERED */;
CREATE TABLE `t11` (`a` INT,`b` VARCHAR(255),PRIMARY KEY(`a`, `b`) /*T![clustered_index] NONCLUSTERED */);
CREATE TABLE `t12` (`id` INT,`a` VARCHAR(255) PRIMARY KEY /*T![clustered_index] CLUSTERED */);
ALTER TABLE `t12` /*T![force_inc] FORCE */ AUTO_INCREMENT = 12;
CREATE TABLE `t13` (`id` VARCHAR(10) PRIMARY KEY,`c1` VARCHAR(10)) ENGINE = InnoDB DEFAULT CHARACTER SET = UTF8MB4 DEFAULT COLLATE = UTF8MB4_BIN /*T! SHARD_ROW_ID_BITS = 4 */ /*T! PRE_SPLIT_REGIONS = 3 */;
```

## tiup playground v4.0.0
```
-> % mysql -uroot -h127.0.0.1 -P4000
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 3
Server version: 5.7.25-TiDB-v4.0.0 TiDB Server (Apache License 2.0) Community Edition, MySQL 5.7 compatible

Copyright (c) 2000, 2019, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> source test.sql;
Query OK, 0 rows affected (0.08 sec)

Database changed
Query OK, 0 rows affected (0.10 sec)

Query OK, 0 rows affected (0.06 sec)

Query OK, 0 rows affected (0.12 sec)

Query OK, 0 rows affected (0.07 sec)

Query OK, 0 rows affected (0.12 sec)

Query OK, 0 rows affected (0.06 sec)

Query OK, 0 rows affected (0.11 sec)

Query OK, 0 rows affected (0.07 sec)

Query OK, 0 rows affected (0.10 sec)

Query OK, 0 rows affected (0.06 sec)

ERROR 8200 (HY000): Unsupported add primary key, alter-primary-key is false
Query OK, 0 rows affected (0.16 sec)

Query OK, 0 rows affected (0.19 sec)

Query OK, 0 rows affected (0.06 sec)

Query OK, 0 rows affected (0.12 sec)`
```

## tiup playground v5.0.0
```
-> % mysql -uroot -h127.0.0.1 -P4000
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 5
Server version: 5.7.25-TiDB-v5.0.0 TiDB Server (Apache License 2.0) Community Edition, MySQL 5.7 compatible

Copyright (c) 2000, 2019, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> source test.sql;
Query OK, 0 rows affected (0.12 sec)

Database changed
Query OK, 0 rows affected (0.07 sec)

Query OK, 0 rows affected (0.13 sec)

Query OK, 0 rows affected (0.07 sec)

Query OK, 0 rows affected (0.11 sec)

Query OK, 0 rows affected (0.07 sec)

Query OK, 0 rows affected (0.12 sec)

Query OK, 0 rows affected (0.06 sec)

Query OK, 0 rows affected (0.13 sec)

Query OK, 0 rows affected (0.06 sec)

Query OK, 0 rows affected (0.10 sec)

Query OK, 0 rows affected (2.92 sec)

Query OK, 0 rows affected (0.15 sec)

Query OK, 0 rows affected (0.07 sec)

Query OK, 0 rows affected (0.15 sec)

Query OK, 0 rows affected (0.07 sec)
```

## tiup playground v5.1.0
```
-> % mysql -uroot -h127.0.0.1 -P4000
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 7
Server version: 5.7.25-TiDB-v5.1.0 TiDB Server (Apache License 2.0) Community Edition, MySQL 5.7 compatible

Copyright (c) 2000, 2019, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> source test.sql;
Query OK, 0 rows affected (0.12 sec)

Database changed
Query OK, 0 rows affected (0.07 sec)

Query OK, 0 rows affected (0.12 sec)

Query OK, 0 rows affected (0.06 sec)

Query OK, 0 rows affected (0.12 sec)

Query OK, 0 rows affected (0.06 sec)

Query OK, 0 rows affected (0.12 sec)

Query OK, 0 rows affected (0.07 sec)

Query OK, 0 rows affected (0.09 sec)

Query OK, 0 rows affected (0.07 sec)

Query OK, 0 rows affected (0.12 sec)

Query OK, 0 rows affected (2.82 sec)

Query OK, 0 rows affected (0.12 sec)

Query OK, 0 rows affected (0.06 sec)

Query OK, 0 rows affected (0.12 sec)

Query OK, 0 rows affected (0.06 sec)
```

## tiup playgound v5.2.0
```
-> % mysql -uroot -h127.0.0.1 -P4000
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 3
Server version: 5.7.25-TiDB-v5.2.0 TiDB Server (Apache License 2.0) Community Edition, MySQL 5.7 compatible

Copyright (c) 2000, 2019, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> source test.sql;
Query OK, 0 rows affected (0.47 sec)

Database changed
Query OK, 0 rows affected (0.18 sec)

Query OK, 0 rows affected (0.19 sec)

Query OK, 0 rows affected (0.18 sec)

Query OK, 0 rows affected (0.18 sec)

Query OK, 0 rows affected (0.18 sec)

Query OK, 0 rows affected (0.19 sec)

Query OK, 0 rows affected (0.21 sec)

Query OK, 0 rows affected (0.07 sec)

Query OK, 0 rows affected (0.14 sec)

Query OK, 0 rows affected (0.06 sec)

Query OK, 0 rows affected (2.91 sec)

Query OK, 0 rows affected (0.06 sec)

Query OK, 0 rows affected (0.12 sec)

Query OK, 0 rows affected (0.06 sec)

Query OK, 0 rows affected (0.11 sec)
```

## tiup playgournd v5.3.0
```
-> % mysql -uroot -h127.0.0.1 -P4000
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 5
Server version: 5.7.25-TiDB-v5.3.0 TiDB Server (Apache License 2.0) Community Edition, MySQL 5.7 compatible

Copyright (c) 2000, 2019, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> source test.sql;
Query OK, 0 rows affected (0.07 sec)

Database changed
Query OK, 0 rows affected (0.13 sec)

Query OK, 0 rows affected (0.07 sec)

Query OK, 0 rows affected (0.12 sec)

Query OK, 0 rows affected (0.06 sec)

Query OK, 0 rows affected (0.17 sec)

Query OK, 0 rows affected (0.06 sec)

Query OK, 0 rows affected (0.13 sec)

Query OK, 0 rows affected (0.09 sec)

Query OK, 0 rows affected (0.95 sec)

Query OK, 0 rows affected (0.19 sec)

Query OK, 0 rows affected (3.01 sec)

Query OK, 0 rows affected (0.13 sec)

Query OK, 0 rows affected (0.07 sec)

Query OK, 0 rows affected (0.10 sec)

Query OK, 0 rows affected (0.06 sec)
```

**Review by overvenus — APPROVED** [review]
LGTM

**Pull request merged**